### PR TITLE
[Feature] Add local tablet reader and rpc interface

### LINKED ISSUE #16535: [Enhancement] Add row store related read interface
## Enhancement

Add row store related interface including TableReader, LocalTabletReader, and also related RPC interfaces.

### PR DESCRIPTION
**Pull request by decster**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #16535

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: gensrc/proto/internal_service.proto [hunk 1]
```diff
@@ -257,6 +257,64 @@ message PTabletWriterCancelResult {
...
 };
 
+message PTabletReaderOpenRequest {
+    optional PUniqueId id = 1;
+    optional int64 tablet_id = 2;
+    optional int64 version = 3;
+}
+
+message PTabletReaderOpenResult {
+    optional StatusPB status = 1;
```
Comment (banmoy) [inline gensrc/proto/internal_service.proto:267]: Why is it not `required` as that in `PTabletReaderMultiGetResult`?

Comment (decster) [reply gensrc/proto/internal_service.proto:267]: ok,  all will change to optional

File: be/src/storage/local_tablet_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+    RETURN_IF_ERROR(_tablet->updates()->get_column_values(read_column_ids, false, rowids_by_rssid, &read_columns));
+
+    // reorder read values to input keys' order and put into values output parameter
+    values.reset();
+    for (size_t col_idx = 0; col_idx < read_column_ids.size(); col_idx++) {
+        values.get_column_by_index(col_idx)->append_selective(*read_columns[col_idx], idxes.data(), 0, idxes.size());
+    }
+    int64_t t_end = MonotonicMillis();
+    LOG(INFO) << strings::Substitute("multi_get tablet:$0 version:$1 #columns:$2 #rows:$3 found:$4 time:$5ms",
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/storage/local_tablet_reader.cpp:146]: will change to VLOG after the implementation is stable.

File: be/src/storage/local_tablet_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        read_column_ids.push_back(cid);
+    }
+
+    // convert keys to pk single column format
+    vector<uint32_t> pk_columns;
+    for (size_t i = 0; i < tablet_schema.num_key_columns(); i++) {
+        pk_columns.push_back((uint32_t)i);
+    }
+    VectorizedSchema pkey_schema = ChunkHelper::convert_schema(tablet_schema, pk_columns);
+    std::unique_ptr<Column> pk_column;
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/storage/local_tablet_reader.cpp:108]: ok

File: gensrc/proto/internal_service.proto (hunk 1 above)
Comment (decster) [reply gensrc/proto/internal_service.proto:267]: according to new rule, all new pb should use optional, required is forbidden, and proto3 has removed required

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/storage/local_tablet_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        pk_columns.push_back((uint32_t)i);
+    }
+    std::unique_ptr<Column> pk_column;
+    if (!PrimaryKeyEncoder::create_column(*tablet_schema.schema(), &pk_column).ok()) {
+        CHECK(false) << "create column for primary key encoder failed";
+    }
+    PrimaryKeyEncoder::encode(*tablet_schema.schema(), keys, 0, keys.num_rows(), pk_column.get());
+
+    // search pks in pk index to get rowids
+    EditVersion editVersion;
```
Comment (mofeiatwork) [inline be/src/storage/local_tablet_reader.cpp:114]: `edit_version` ?

File: be/src/storage/table_reader.h [hunk 5]
```diff
@@ -24,7 +24,15 @@
 
 namespace starrocks {
 
-// Parameters used to create a TableReader.
+class LocalTabletReader;
+
+// Parameters used to create a TableReader supporting only single local tablet access
+struct LocalTableReaderParams {
+    int64_t tablet_id;
+    uint64_t version;
```
Comment (mofeiatwork) [inline be/src/storage/table_reader.h:32]: `int64_t` ?

File: gensrc/proto/doris_internal_service.proto [hunk 6]
```diff
@@ -66,4 +66,11 @@ service PBackendService {
 
     // NOTE(murphy) check if it's really necessary
     rpc submit_mv_maintenance_task(starrocks.PMVMaintenanceTaskRequest) returns (starrocks.PMVMaintenanceTaskResult);
+
+    // local tablet reader
+    rpc local_tablet_reader_open(starrocks.PTabletReaderOpenRequest) returns (starrocks.PTabletReaderOpenResult);
```
Comment (mofeiatwork) [inline gensrc/proto/doris_internal_service.proto:71]: What if opened and not closed? 
If there any configuration about timeout or session ?

Comment (decster) [reply gensrc/proto/doris_internal_service.proto:71]: There will be a localtabletmanager to manage all opened readers and do the GC work. yes we may need to add more params like timeout/sessionid(that's a larger scope, and we haven't discussed it yet) so maybe not included in this PR?

File: be/src/storage/table_reader.h (hunk 5 above)
Comment (decster) [reply be/src/storage/table_reader.h:32]: fixed

File: be/src/storage/local_tablet_reader.cpp (hunk 4 above)
Comment (decster) [reply be/src/storage/local_tablet_reader.cpp:114]: fixed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by banmoy — DISMISSED** [review]

**Comment by banmoy** [comment]
@decster starrocks_clang-format and starrocks_tscancode failed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by decster** [comment]
run starrocks_fe_unittest

**Review by chaoyli — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by chaoyli — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**